File: FlowCrypt/src/main/java/com/flowcrypt/email/model/KeysStorage.java
/*
 * © 2016-2019 FlowCrypt Limited. Limitations apply. Contact human@flowcrypt.com
 * Contributors: DenBond7
 */

package com.flowcrypt.email.model;

import android.content.Context;

import java.util.List;

import androidx.annotation.Keep;

@Keep
public interface KeysStorage {

  PgpContact findPgpContact(String longid);

  // if two contacts requested and only one found, will still return list of 2:
  // eg [PgpContact, null] or [null, PgpContact] depending which one is missing
  List findPgpContacts(String[] longid);

  PgpKeyInfo getPgpPrivateKey(String longid);

  // if 2 keys requested and only one found, will return list of 1: [PgpKey]
  List<PgpKeyInfo> getFilteredPgpPrivateKeys(String[] longid);

  List<PgpKeyInfo> getAllPgpPrivateKeys();

  String getPassphrase(String longid);

  void refresh(Context context);

}

File: FlowCrypt/src/main/java/com/flowcrypt/email/model/KeysStorage.kt
/*
 * © 2016-present FlowCrypt a.s. Limitations apply. Contact human@flowcrypt.com
 * Contributors: DenBond7
 */

package com.flowcrypt.email.model

import androidx.annotation.Keep
import com.flowcrypt.email.database.entity.KeyEntity
import com.flowcrypt.email.security.model.PgpKeyDetails
import kotlinx.coroutines.flow.Flow
import org.bouncycastle.openpgp.PGPSecretKeyRing
import org.pgpainless.key.protection.SecretKeyRingProtector
import org.pgpainless.util.Passphrase
import java.time.Instant

@Keep
interface KeysStorage {
  fun getRawKeys(): List<KeyEntity>

  fun getPGPSecretKeyRings(): List<PGPSecretKeyRing>

  fun getPgpKeyDetailsList(): List<PgpKeyDetails>

  fun getPGPSecretKeyRingByFingerprint(fingerprint: String): PGPSecretKeyRing?

  fun getPGPSecretKeyRingsByFingerprints(fingerprints: Collection<String>): List<PGPSecretKeyRing>

  fun getPGPSecretKeyRingsByUserId(user: String): List<PGPSecretKeyRing>

  fun getPassphraseByFingerprint(fingerprint: String): Passphrase?

  fun getPassphraseTypeByFingerprint(fingerprint: String): KeyEntity.PassphraseType?

  fun getSecretKeyRingProtector(): SecretKeyRingProtector

  fun updatePassphrasesCache()

  fun clearPassphrasesCache()

  fun putPassphraseToCache(
    fingerprint: String,
    passphrase: Passphrase,
    validUntil: Instant,
    passphraseType: KeyEntity.PassphraseType
  )

  fun hasEmptyPassphrase(vararg types: KeyEntity.PassphraseType): Boolean

  fun hasNonEmptyPassphrase(vararg types: KeyEntity.PassphraseType): Boolean

  fun hasPassphrase(passphrase: Passphrase): Boolean

  fun getFingerprintsWithEmptyPassphrase(): List<String>

  fun getFirstUsableForEncryptionPGPSecretKeyRing(user: String): PGPSecretKeyRing?

  fun getPassPhrasesUpdatesFlow(): Flow<Long>
}
